src/Sylius/Bundle/ApiBundle/Validator/CatalogPromotion/ForTaxonsScopeValidator.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Validator\CatalogPromotion;
+
+use Sylius\Bundle\ApiBundle\SectionResolver\AdminApiSection;
+use Sylius\Bundle\CoreBundle\SectionResolver\SectionProviderInterface;
+use Sylius\Bundle\CoreBundle\Validator\CatalogPromotionScope\ScopeValidatorInterface;
+use Sylius\Bundle\CoreBundle\Validator\Constraints\CatalogPromotionScope;
+use Sylius\Component\Taxonomy\Repository\TaxonRepositoryInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\Context\ExecutionContextInterface;
+use Webmozart\Assert\Assert;
+
+final class ForTaxonsScopeValidator implements ScopeValidatorInterface
+{
+    public function __construct(
+        private ScopeValidatorInterface $baseScopeValidator,
+        private SectionProviderInterface $sectionProvider
+    ) {
+    }
+
+    public function validate(array $configuration, Constraint $constraint, ExecutionContextInterface $context): void
+    {
+        if (!$this->sectionProvider->getSection() instanceof AdminApiSection) {
+            $this->baseScopeValidator->validate($configuration, $constraint, $context);
+
+            return;
+        }
+
+        /** @var CatalogPromotionScope $constraint */
+        Assert::isInstanceOf($constraint, CatalogPromotionScope::class);
+
+        if (!isset($configuration['taxons']) || empty($configuration['taxons'])) {
+            $context->buildViolation($constraint->taxonsNotEmpty)->atPath('configuration.taxons')->addViolation();
+
+            return;
+        }
+
+        $this->baseScopeValidator->validate($configuration, $constraint, $context);
```

Comment:
What do you think about reversing execution to have only one base execution?

```suggestion
        /** @var CatalogPromotionScope $constraint */
        Assert::isInstanceOf($constraint, CatalogPromotionScope::class);
        
        if ($this->sectionProvider->getSection() instanceof AdminApiSection) {
            if (!isset($configuration['taxons']) || empty($configuration['taxons'])) {
                $context->buildViolation($constraint->taxonsNotEmpty)->atPath('configuration.taxons')->addViolation();

                return;
            }
        }

        $this->baseScopeValidator->validate($configuration, $constraint, $context);
```
or to be more precise
```suggestion
        /** @var CatalogPromotionScope $constraint */
        Assert::isInstanceOf($constraint, CatalogPromotionScope::class);
        
        if ($this->sectionProvider->getSection() instanceof AdminApiSection && $this->isTaxonMissingInConfiguration($configuration)) {
            $context->buildViolation($constraint->taxonsNotEmpty)->atPath('configuration.taxons')->addViolation();

            return;
        }

        $this->baseScopeValidator->validate($configuration, $constraint, $context);
```

Reply:
If you like it and decide to use this structure, same may be done in `ForProductsScopeValidator` and `ForVariantsScopeValidator`

Reply:
👍